lib/src/Uncertainty/Algorithm/WeightedExperiments/TensorProductExperiment.cxx

```diff
@@ -0,0 +1,172 @@
+  if (!(numberOfMarginalExperiments > 0))
+    throw InvalidArgumentException(HERE) << "Error: expected a positive number of marginal experiments, here it is " << numberOfMarginalExperiments;
+  // Compute marginal nodes and weights
+  LOGDEBUG(OSS() << "+ Compute marginal nodes and weights");
+  Collection<Sample> marginalNodesCollection(numberOfMarginalExperiments);
+  Collection<Point> marginalWeightsCollection(numberOfMarginalExperiments);
+  for (UnsignedInteger i = 0; i < numberOfMarginalExperiments; ++i)
+  {
+    LOGDEBUG(OSS() << "  Generate weights and nodes for marginal experiment = " << i);
+    Point weights;
+    LOGDEBUG(OSS() << "  collection_[i] = " << collection_[i]);
+    const Sample nodes(collection_[i].generateWithWeights(weights));
+    marginalNodesCollection[i] = nodes;
+    marginalWeightsCollection[i] = weights;
+    LOGDEBUG(OSS() << "  weights = " << weights);
+    LOGDEBUG(OSS() << "  nodes = ");
+    LOGDEBUG(OSS() << nodes);
+  }
+  // Compute dimension
+  UnsignedInteger dimension = 0;
+  for (UnsignedInteger i = 0; i < numberOfMarginalExperiments; ++i)
+  {
+    dimension += marginalNodesCollection[i].getDimension();
+  }
+  LOGDEBUG(OSS() << "+ dimension = " << dimension);
+  // Compute size for each marginal experiment.
+  LOGDEBUG(OSS() << "+ Compute size for each dimension");
+  Indices marginalSizeIndices(numberOfMarginalExperiments);
+  for (UnsignedInteger i = 0; i < numberOfMarginalExperiments; ++i)
+  {
+    LOGDEBUG(OSS() << "  marginalWeightsCollection[i] = " << marginalWeightsCollection[i]);
+    marginalSizeIndices[i] = marginalWeightsCollection[i].getDimension();
+    LOGDEBUG(OSS() << "  marginalSizeIndices[i] = " << marginalSizeIndices[i]);
+  }
+  LOGDEBUG(OSS() << "  marginalSizeIndices = " << marginalSizeIndices);
+  // Create index set
+  LOGDEBUG(OSS() << "+ Create index set");
+  Tuples tuples(marginalSizeIndices);
+  const IndicesCollection indicesCollection(tuples.generate());
+  // Compute nodes and weights
+  LOGDEBUG(OSS() << "+ Compute nodes and weights");
+  const UnsignedInteger size = indicesCollection.getSize();
+  LOGDEBUG(OSS() << "  Total size = " << size);
+  weights_ = Point(size);
+  nodes_ = Sample(size, dimension);
+  for (UnsignedInteger i = 0; i < size; ++i)
+  {
+    weights_[i] = 1.0;
+    UnsignedInteger marginal_experiment_index = 0;
+    for (UnsignedInteger j = 0; j < numberOfMarginalExperiments; ++j)
+    {
+      const UnsignedInteger marginal_index = indicesCollection(i, j);
+      const Point marginalWeights(marginalWeightsCollection[j]);
+      weights_[i] *= marginalWeights[marginal_index];
+      const Sample marginalNodes(marginalNodesCollection[j]);
+      UnsignedInteger marginal_dimension = marginalNodes.getDimension();
+      for (UnsignedInteger k = 0; k < marginal_dimension; ++k)
+      {
+        nodes_(i, marginal_experiment_index) = marginalNodes(marginal_index, k);
+        marginal_experiment_index += 1;
```

Comment:
++marginalExperimentIndex;

Reply:
Ok. Fixed. Also renamed it to camelCase.